src/metabase/util/queue.clj

```diff
@@ -84,3 +88,94 @@
   ([^DelayQueue queue max-items ^long max-first-ms ^long max-next-ms]
    (when-let [fst (.poll queue max-first-ms TimeUnit/MILLISECONDS)]
      (take-delayed-batch* queue max-items max-next-ms [(:value fst)]))))
+
+(defonce ^:private listeners (atom {}))
+
+(defn- listener-thread [{:keys [listener-name queue handler result-handler err-handler finally-handler max-batch-items max-next-ms]}]
```

Comment:
A concern I have with exposing this very general interface, is that it's calling down into code that assumes a very specialized `DelayQueue`. Looking at `take-delayed-batch!` and its helper though, I think you can relax it to just assume a `BlockingQueue`, so let's do that, and make sure all the docstrings reflect that assumption.

Reply:
Another option is just to defer generalizing until we have more examples, this might be premature.